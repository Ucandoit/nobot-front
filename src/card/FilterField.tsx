import { FormControl, InputLabel, makeStyles, MenuItem, Select } from '@material-ui/core';
import React, { ChangeEvent, useCallback } from 'react';
import { Card, CardFilters } from '../helpers';

const useStyles = makeStyles(theme => ({
  filterField: {
    margin: theme.spacing(1),
    minWidth: '10rem'
  },
  select: {
    paddingTop: '1rem',
    paddingBottom: '1rem'
  }
}));

interface FilterFieldProps {
  property: keyof Card;
  label: string;
  items: (string | number)[];
  value: string | number | undefined;
  changeFilter: (property: keyof CardFilters, value: string) => void;
}

const FilterField = ({ property, label, items, value, changeFilter }: FilterFieldProps) => {
  const classes = useStyles();
  const onChange = useCallback(
    (event: ChangeEvent<{ value: unknown }>) =>
      changeFilter(property as keyof CardFilters, event.target.value as string),
    [property, changeFilter]
  );
  return (
    <FormControl variant="outlined" margin="dense" className={classes.filterField}>
      <InputLabel id={`${property}-label`}>{label}</InputLabel>
      <Select labelId={`${property}-label`} label={label} value={value || ''} onChange={onChange}>
        <MenuItem value="">-</MenuItem>
        {items.map(item => (
          <MenuItem key={item} value={item}>
            {item}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

export default FilterField;
