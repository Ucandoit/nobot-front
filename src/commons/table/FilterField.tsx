import { FormControl, InputLabel, makeStyles, MenuItem, Select } from '@material-ui/core';
import React, { ChangeEvent, useCallback } from 'react';
import { FilterType } from './types';

const useStyles = makeStyles(theme => ({
  filterField: {
    margin: theme.spacing(1),
    minWidth: '10rem'
  },
  select: {
    paddingTop: '1rem',
    paddingBottom: '1rem'
  }
}));

interface FilterFieldProps<T> {
  property: keyof T;
  label: string;
  type: FilterType;
  items?: Array<string | { key: string; value: string }>;
  value: string | undefined;
  changeFilter: (property: keyof T, value: string) => void;
}
const FilterField = <T,>({ property, label, type, items, value, changeFilter }: FilterFieldProps<T>) => {
  const classes = useStyles();
  const onChange = useCallback(
    (event: ChangeEvent<{ value: unknown }>) => changeFilter(property as keyof T, event.target.value as string),
    [property, changeFilter]
  );
  return (
    <FormControl variant="outlined" margin="dense" className={classes.filterField}>
      <InputLabel id={`${property}-label`}>{label}</InputLabel>
      {type === 'select' ? (
        <Select labelId={`${property}-label`} label={label} value={value || ''} onChange={onChange}>
          <MenuItem value="">-</MenuItem>
          {items &&
            items.map((item: string | { key: string; value: string }) => (
              <MenuItem
                key={typeof item === 'object' ? item.key : item}
                value={typeof item === 'object' ? item.value : item}
              >
                {item}
              </MenuItem>
            ))}
        </Select>
      ) : null}
    </FormControl>
  );
};

export default FilterField;
