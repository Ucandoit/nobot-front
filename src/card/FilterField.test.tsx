import { Select } from '@material-ui/core';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import FilterField from './FilterField';

const changeFilter = jest.fn();

describe('FilterField', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    wrapper = shallow(
      <FilterField property="cost" label="Cost" items={[1.0, 1.5, 2, 2.5]} value={''} changeFilter={changeFilter} />
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it(`should render correctly when passing value other than ''`, () => {
    wrapper.setProps({
      value: 1.5
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should call changeFilter function on select', () => {
    wrapper.find(Select).simulate('change', { target: { value: 1.5 } });
    expect(changeFilter).toHaveBeenCalledTimes(1);
    expect(changeFilter).toHaveBeenCalledWith('cost', 1.5);
  });
});
