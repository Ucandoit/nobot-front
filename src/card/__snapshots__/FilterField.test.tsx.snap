// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterField should render correctly 1`] = `
<WithStyles(ForwardRef(FormControl))
  className="makeStyles-filterField-1"
  margin="dense"
  variant="outlined"
>
  <WithStyles(ForwardRef(InputLabel))
    id="cost-label"
  >
    Cost
  </WithStyles(ForwardRef(InputLabel))>
  <WithStyles(ForwardRef(Select))
    label="Cost"
    labelId="cost-label"
    onChange={[Function]}
    value=""
  >
    <WithStyles(ForwardRef(MenuItem))
      value=""
    >
      -
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="1"
      value={1}
    >
      1
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="1.5"
      value={1.5}
    >
      1.5
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="2"
      value={2}
    >
      2
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="2.5"
      value={2.5}
    >
      2.5
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Select))>
</WithStyles(ForwardRef(FormControl))>
`;

exports[`FilterField should render correctly when passing value other than '' 1`] = `
<WithStyles(ForwardRef(FormControl))
  className="makeStyles-filterField-1"
  margin="dense"
  variant="outlined"
>
  <WithStyles(ForwardRef(InputLabel))
    id="cost-label"
  >
    Cost
  </WithStyles(ForwardRef(InputLabel))>
  <WithStyles(ForwardRef(Select))
    label="Cost"
    labelId="cost-label"
    onChange={[Function]}
    value={1.5}
  >
    <WithStyles(ForwardRef(MenuItem))
      value=""
    >
      -
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="1"
      value={1}
    >
      1
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="1.5"
      value={1.5}
    >
      1.5
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="2"
      value={2}
    >
      2
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="2.5"
      value={2.5}
    >
      2.5
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Select))>
</WithStyles(ForwardRef(FormControl))>
`;
